### DISCUSSION TITLE
Security considerations

### DISCUSSION THREAD

**Discussion by stanislaw** [textX/textX#422] [General]
Dear developers of textX/Arpeggio,

I have been using both projects successfully more than 3 years in my https://github.com/strictdoc-project/strictdoc project. Recently I have started taking into account the security aspect of my tool given that it is (slowly) moving towards being deployable in the public networks. 

StrictDoc uses textX/Arpeggio to parse the documentation files written using strict-doc (*.sdoc) markup for which I designed a custom grammar. The obvious security-related questions that come to my mind as I keep using textX/Arpeggio are as follows:

- Does textX/Arpeggio provide any security-related guarantees?
- Is there anything that a user has to know when designing a textX grammar?

Honestly speaking, I didn't spend too much time on this, and I am not a security expert, so in my naive understanding, it would be great to know that at least:

- It is not possible to segfault textX/Arpeggio, that error reporting stays within what is designed in textX/Arpeggio.
- It is not possible to halt the parser. I know it is quite easy if mess up my grammar, however **I think** I made it stable over the years. It would be nice to get some advice here.

An extra observation: textX/Arpeggio do not seem to depend on anything else, so the evaluation of these library's security aspect does not propagate to any upstream libraries recursively which is great.

If I was to answer my questions myself, this is what I would do for StrictDoc:

- Use a fuzzing tool to stress-test my textX parser, feeding all sorts of random inputs, including those that are mutated versions of SDoc documents which I would use as seed.
- Find a way to scan my grammar against the use of unsafe regular expressions and if they exist, eliminate them.
- Is there anything else that you could recommend?

Thank you for creating this tool and I look forward to your reply.

Stanislav

**Comment by igordejanovic** [comment]
Hi @stanislaw. Thanks for bringing this important issue up.

>     * Does textX/Arpeggio provide any security-related guarantees?

Since the maintainers are not security experts we do not provide any security-related guarantees. There has been no security audit of Arpeggio/textX AFAIK.

It would be great if some member(s) from the community with a security background could do the code review.

 >     * Is there anything that a user has to know when designing a textX grammar?

What comes to mind is a possibility to trigger [exponential bahaviour of recursive descent parsing](https://en.wikipedia.org/wiki/Parsing_expression_grammar#Implementing_parsers_from_parsing_expression_grammars).

Try for example this grammar (an example from [this SO post](https://cstheory.stackexchange.com/questions/22520/example-of-context-free-grammar-that-triggers-exponential-behaviour-without-memo) modified to accout for ordered choice in PEG:

```bnf
S: X "b" | X "c";
X: "a" X X | "a" X | "a";
```

with input of the form `a+c` (e.g. `aaaaaaac`) and try to increase the number of `a`s. You will see that parsing time raises quickly and soon the parser will hang in a futile effort to recognize the input.

Now, if attackers could know for this weakness in your grammar they could craft the input in such a way to trigger this behavior and execute [resource exhaustion attack](https://en.wikipedia.org/wiki/Resource_exhaustion_attack).
A very similar problem can be found in regex engines as you already expected.

A way to deal with this exponential behavior is the memoization technique which is supported by Arpeggio but is disabled by default as it incurs a fixed overhead.
The idea is to cache sub-results of recursive calls and reuse it. This bring exponential time to linear albeit at the expense of caching overhead.

Thus, for the above grammar:

```python
mm = metamodel_from_str(grammar, memoization=True)
mm.model_from_file('aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaac')
```

finishes without problem.


>     * It is not possible to segfault textX/Arpeggio, that error reporting stays within what is designed in textX/Arpeggio.

Not that I'm aware of but no guarantees.

>     * It is not possible to halt the parser. I know it is quite easy if mess up my grammar, however **I think** I made it stable over the years. It would be nice to get some advice here.

Yes, if the grammar/regexes have flaws that lead to possibility of resource exhaustion.
Halting is also possible if the grammar has left recursive rules, whether direct or indirect (or hidden). But this is usually easy to detect by a decent test coverage.

> If I was to answer my questions myself, this is what I would do for StrictDoc:
> 
>     * Use a fuzzing tool to stress-test my textX parser, feeding all sorts of random inputs, including those that are mutated versions of SDoc documents which I would use as seed.
> 
>     * Find a way to scan my grammar against the use of unsafe regular expressions and if they exist, eliminate them.
> 
>     * Is there anything else that you could recommend?

This pretty much covers what I would consider. Stress testing and scanning for pathological regexes that could trigger exponential behavior should cover most of the concerns. Together, with activating memoization just to be on the safe side, if you can take the incurred overhead.

Some notes on memoization:
- You should test your grammar with/without memoization. It might be that memoization will bring performance improvement if you already have some rules that trigger super-polinomial behaviour.
- I haven't checked if memoization itself can be used for memory exhaustion attack. I think that the memory consumtion complexity is linear with respect to the input so it shouldn't be the case.
